mkosi/__init__.py

```diff
@@ -1545,7 +1546,50 @@ def build_initrd(context: Context) -> Path:
     return config.output_dir / config.output
 
 
+def identify_cpu(root: Path) -> tuple[Optional[Path], Optional[Path]]:
+    for entry in Path("/proc/cpuinfo").read_text().split("\n\n"):
+        vendor_id = family = model = stepping = None
+        for line in entry.splitlines():
+            key, _, value = line.partition(":")
+            key = key.strip()
+            value = value.strip()
+
+            if key == "vendor_id":
+                vendor_id = value
+            elif key == "cpu family":
+                family = int(value)
+            elif key == "model":
+                model = int(value)
+            elif key == "stepping":
+                stepping = int(value)
+
+        if vendor_id is not None and family is not None and model is not None and stepping is not None:
+            break
+    else:
+        return (None, None)
+
+    if vendor_id == "AuthenticAMD":
+        uroot = root / "usr/lib/firmware/amd-ucode"
+        if family > 21:
+            ucode = uroot / f"microcode_amd_fam{family:x}h.bin"
+        else:
+            ucode = uroot / "microcode_amd.bin"
+        if ucode.exists():
+            return (Path(f"{vendor_id}.bin"), ucode)
+    elif vendor_id == "GenuineIntel":
+        uroot = root / "usr/lib/firmware/intel-ucode"
+        if (ucode := (uroot / f"{family:02x}-{model:02x}-{stepping:02x}")).exists():
+            return (Path(f"{vendor_id}.bin"), ucode)
+        if (ucode := (uroot / f"{family:02x}-{model:02x}-{stepping:02x}.initramfs")).exists():
+            return (Path(f"{vendor_id}.bin"), ucode)
+
+    return (Path(f"{vendor_id}.bin"), None)
+
+
 def build_microcode_initrd(context: Context) -> Optional[Path]:
+    if context.config.architecture not in [Architecture.x86, Architecture.x86_64]:
```

Comment:
Use tuple instead of list?

Reply:
It doesn't make much of a difference under the hood - the list can't really be changed - and I prefer the visual distinctiveness of square braces. With parens I'm always thinking twice whether it's a tuple or a parenthesised expression.

Reply:
The thing is we use parentheses almost everywhere else, so this breaks from the established practice

Reply:
Changed it for uniformities sake (though there are some differing places still), but generally I think square braces would be more readable and tuples don't have any real upside in these positions.